### ISSUE TITLE
Support direct download where file size is too large for homeserver

### ISSUE THREAD

**Issue by 0x1a8510f2** [vector-im/element-android#3274]
#### Is your feature request related to a problem? Please describe.
As per [this](https://matrix.to/#/!hFdiidJUHZOsUlmhcn:matrix.org/$v5yKMoqeP3bs5yRdfAZ-1FBQyiTS5IUd4E1rDMtLSP8?via=an-atom-in.space&via=matrix.org&via=tchncs.de) conversation, sometimes your homeserver has a file size limit lower than someone else's, but you still wish to receive a file they sent. To do this, you can manually download directly from their homeserver, but it would be great if Element could instead detect that the file is too large for your homeserver and offer an automatic direct download from the origin homeserver instead (in fact, for any file federation failures).

#### Describe the solution you'd like.
I know it's possible to get a homeserver's max upload limit, and also media/file events have a "size" key. Hence, this information can be used to determine if the file will be downloaded by the homeserver. In cases where it is not downloaded by the homeserver due to the limit, but where the user explicitly requested to download a file (for example, using the "save" option), Element could show a dialogue asking whether the file should be fetched directly from the origin homeserver and warning that this will disclose the user's IP to that homeserver. This behaviour could also be editable in settings with options like "prompt" (default), "always yes" and "always no".

This can also be extended further to any file federation errors, although this will require some consideration as not all download errors should trigger this (for instance, a connection timeout due to lack of internet connection should not trigger this). This would additionally help in situations where the "size" key is wrong.

#### Describe alternatives you've considered.
- Manual downloading - while possible, it's quite inconvenient and some users would not know how to do this

#### Additional context
- [Aforementioned conversation leading to this idea](https://matrix.to/#/!hFdiidJUHZOsUlmhcn:matrix.org/$v5yKMoqeP3bs5yRdfAZ-1FBQyiTS5IUd4E1rDMtLSP8?via=an-atom-in.space&via=matrix.org&via=tchncs.de)

**Comment by bmarty** [comment]
I'm not a specialist of the media repository, nor of how the federation work for attachment, but to me the size limit occurs only when you upload a file because the file will be stored on your homeserver and it's not free. When downloading a file, you request the file from your homeserver, and the homeserver will get it from the origin server, if necessary, where it has been uploaded. At least this is how I would have specced it, to limit the duplication of attachments to every homeservers.

**Comment by 0x1a8510f2** [comment]
@bmarty that's all correct, except when the file is too large for your homeserver, it will not be downloaded over federation, which means that the limit still applies when downloading from any homeserver other than your own. That's why I'm suggesting to offer an option to download directly from the homeserver/media repo where the file was originally uploaded.
